### PR TITLE
Implement safe_ptr for safely accessing unowned objects

### PR DESCRIPTION
**Pull request by felipemanga**
Since existing code passes raw pointers around, it is not always possible to know if a pointer to an object is still valid or not.
Generally, the standard solution would be to use shared_ptr together with weak_ptr, but that would involve a large rewrite.
As a workaround, this PR contains a smart pointer (this_ptr) that automatically gets set to nullptr when the initial this_ptr it was copied from is destroyed. 
In short, it gets rid of dangling-pointer problems.
Please see comment in header for further clarification/examples.

### CODE REVIEW

**Review by Zughy — APPROVED** [review]
needed band-aid
